### PR TITLE
mautrix-telegram: patch away alembic dependency

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

`alembic`[1] is a database migration tool which is invoked from the CLI
when installing the telegram bridge, but never needed during the
runtime.

The reason why `alembic` is required here is to ensure that it
exists in the Python environment when deploying the bridge. However
`alembic` requires `mautrix-telegram` in its environment to create a
database schema from the Python models.

Such a dependency relation may be possible with tools like virtualenv,
however it'll result in an infinite recursion at evaluation time in Nix.

With this patch, `mautrix-telegram` doesn't depend on `alembic` anymore
and provides a patched alembic (`pkgs.mautrix-telegram.alembic`) which
has `mautrix-telegram` in its path.

[1] https://alembic.sqlalchemy.org/en/latest/

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
As the package maintainer gave a :+1: I guess this should be fine. I'll merge this PR in a week unless there are further comments :)

**Review by worldofpeace — APPROVED** [review]
I recall reviewing this when it was added to nixpkgs and maybe suggesting something like this being done (or perhaps thought). LGTM.

Perhaps initial-setup should be tested though.

**Comment by Ma27** [comment]
@worldofpeace I have to admit that I didn't read the thread when this was packaged initially, this is actually the result of something I used to have in an overlay in my deployment repo that seemed useful to upstream :)

> Perhaps initial-setup should be tested though.

I actually confirmed that `mautrix-telegram.alembic` is able ot create a sqlite database that is usable by `mautrix-telegram`. Is there anything else you'd suggest to test?

**Comment by worldofpeace** [comment]
> I actually confirmed that mautrix-telegram.alembic is able ot create a sqlite database that is usable by mautrix-telegram. Is there anything else you'd suggest to test?

I think doing `alembic upgrade head` would be the only thing to test. 

But now I'm thinking, how is a user going to know that they need to use `mautrix-telegram.alembic`?
Also will hydra cache `mautrix-telegram.alembic`?

**Comment by Ma27** [comment]
> But now I'm thinking, how is a user going to know that they need to use mautrix-telegram.alembic

Whenever you need `alembic` to run migrations for `mautrix-telegram` as this should be the only case where you need `alembic` with a dependency to the `mautrix-telegram` package, right?

> Also will hydra cache mautrix-telegram.alembic?

That's indeed a good question. I guess not (atm), but this should be fixable by `recurseIntoAttrs`, right?

**Comment by worldofpeace** [comment]
> That's indeed a good question. I guess not (atm), but this should be fixable by recurseIntoAttrs, right?

I believe that should be correct.

**Comment by Ma27** [comment]
@worldofpeace ack, fixed :)

**Comment by Ma27** [comment]
As announced, I'll merge this now. It's been approved by a collaborator and got a thumbs-up from the  package maintainer. Also, `recurseIntoAttrs` is now used to get a prebuilt `mautrix-telegram.alembic` from Hydra.

**Comment by worldofpeace** [comment]
Thanks for being so thorough :sparkles:

**Pull request merged**
